### ISSUE TITLE
mesh-expansion: reboot deletes required directory, sidecar startup fails

### ISSUE THREAD

**Issue by Christoph-Raab** [istio/istio#42603]
### Bug Description

When a mesh-expansion vm with a sidecar reboots, the `/var/run/secrets` directory is deleted. As discussed in https://github.com/istio/istio/issues/35343 this doesn't seam to be a problem regarding the `tokens/` subdirectory.

But with istio 1.14 the deletion leads to a failing `SDS`, because the sidecar wants to recreate `/var/run/secrets` to create `/var/run/secrets/workload-spiffe-uds/socket`. The directory creation fails with `permission denied`. Because of this the sidecar isn't able to come up.

Our puppet setup would recreate the directory, but since we're using istio's dns proxying, the puppet agent can not run, because dns isn't working without a running sidecar.

I can think of a workaround to recreate the directory after a reboot, but I'd argue that shouldn't be required for the package to be used. I understand that the content of the directory should be purged, but the directory itself shouldn't be deleted if it's required.

Also I would suggest to add this behavior to the docs page, took some time to figure out what the problem was.

### Version

```prose
$ istioctl version
client version: 1.14.5
control plane version: 1.14.5
data plane version: 1.14.5 (89 proxies)

$ uname -a
Linux <hostname> 5.10.0-19-amd64 #1 SMP Debian 5.10.149-2 (2022-10-21) x86_64 GNU/Linux
```

### Additional Information

2022-12-29T08:08:51.616541Z	info	JWT policy is third-party-jwt
2022-12-29T08:08:51.616545Z	info	using credential fetcher of JWT type in cluster.local trust domain
2022-12-29T08:08:51.632722Z	info	dns	Starting local udp DNS server on 127.0.0.1:15053
2022-12-29T08:08:51.633015Z	info	dns	Starting local tcp DNS server on 127.0.0.1:15053
2022-12-29T08:08:51.633040Z	info	SDS socket not found. Starting Istio SDS Server
2022-12-29T08:08:51.633048Z	info	CA Endpoint istiod.istio-system.svc:15012, provider Citadel
2022-12-29T08:08:51.633561Z	info	Using CA istiod.istio-system.svc:15012 cert with certs: ./etc/certs/root-cert.pem
2022-12-29T08:08:51.634496Z	info	citadelclient	Citadel client using custom root cert: ./etc/certs/root-cert.pem
2022-12-29T08:08:51.652824Z	info	Opening status port 15020
2022-12-29T08:08:51.956885Z	info	ads	All caches have been synced up in 368.392041ms, marking server ready
**2022-12-29T08:08:51.957100Z	warn	Failed to create directory for ./var/run/secrets/workload-spiffe-uds/socket: mkdir var/run/secrets: permission denied
2022-12-29T08:08:51.957140Z	error	sds	Failed to set up UDS path: failed to listen on unix socket "./var/run/secrets/workload-spiffe-uds/socket": listen unix ./var/run/secrets/workload-spiffe-uds/socket: bind: no such file or directory
2022-12-29T08:08:51.957145Z	info	sds	SDS server for workload certificates started, listening on "./var/run/secrets/workload-spiffe-uds/socket"**
2022-12-29T08:08:51.957175Z	info	xdsproxy	Initializing with upstream address "istiod.istio-system.svc:15012" and cluster "kube-cluster"
2022-12-29T08:08:51.958777Z	info	starting Http service at 127.0.0.1:15004
2022-12-29T08:08:51.979122Z	info	sds	Starting SDS grpc server
2022-12-29T08:08:51.979183Z	warn	Failed to create directory for ./var/run/secrets/workload-spiffe-uds/socket: mkdir var/run/secrets: permission denied
2022-12-29T08:08:51.979221Z	error	sds	SDS grpc server for workload proxies failed to set up UDS: failed to listen on unix socket "./var/run/secrets/workload-spiffe-uds/socket": listen unix ./var/run/secrets/workload-spiffe-uds/socket: bind: no such file or directory
2022-12-29T08:08:51.979409Z	info	Pilot SAN: [istiod.istio-system.svc]
2022-12-29T08:08:51.981411Z	info	Starting proxy agent
2022-12-29T08:08:51.981436Z	info	Epoch 0 starting
2022-12-29T08:08:51.981457Z	info	Envoy command: [-c etc/istio/proxy/envoy-rev0.json --restart-epoch 0 --drain-time-s 45 --drain-strategy immediate --parent-shutdown-time-s 60 --local-address-ip-version v4 --file-flush-interval-msec 1000 --disable-hot-restart --log-format %Y-%m-%dT%T.%fZ	%l	envoy %n	%v -l warning --component-log-level misc:error --concurrency 2]
2022-12-29T08:08:52.867830Z	warn	failed to fetch token from file: open ./var/run/secrets/tokens/istio-token: no such file or directory
2022-12-29T08:08:52.867885Z	info	xdsproxy	connected to upstream XDS server: istiod.istio-system.svc:15012
2022-12-29T08:08:52.988109Z	warn	Failed to create directory for ./var/run/secrets/workload-spiffe-uds/socket: mkdir var/run/secrets: permission denied
2022-12-29T08:08:52.988236Z	error	sds	SDS grpc server for workload proxies failed to set up UDS: failed to listen on unix socket "./var/run/secrets/workload-spiffe-uds/socket": listen unix ./var/run/secrets/workload-spiffe-uds/socket: bind: no such file or directory
2022-12-29T08:08:53.359444Z	warn	failed to fetch token from file: open ./var/run/secrets/tokens/istio-token: no such file or directory
2022-12-29T08:08:53.398696Z	info	cache	generated new workload certificate	latency=1.439793431s ttl=23h59m57.601313949s
2022-12-29T08:08:53.398741Z	info	cache	Root cert has changed, start rotating root cert
2022-12-29T08:08:53.398776Z	info	ads	XDS: Incremental Pushing:0 ConnectedEndpoints:0 Version:
2022-12-29T08:08:53.403606Z	info	cache	returned workload trust anchor from cache	ttl=23h59m57.596400108s
2022-12-29T08:08:54.988727Z	warn	Failed to create directory for ./var/run/secrets/workload-spiffe-uds/socket: mkdir var/run/secrets: permission denied
2022-12-29T08:08:54.988931Z	error	sds	SDS grpc server for workload proxies failed to set up UDS: failed to listen on unix socket "./var/run/secrets/workload-spiffe-uds/socket": listen unix ./var/run/secrets/workload-spiffe-uds/socket: bind: no such file or directory
2022-12-29T08:08:57.521451Z	warn	Failed to create directory for ./var/run/secrets/workload-spiffe-uds/socket: mkdir var/run/secrets: permission denied
2022-12-29T08:08:57.521525Z	error	sds	SDS grpc server for workload proxies failed to set up UDS: failed to listen on unix socket "./var/run/secrets/workload-spiffe-uds/socket": listen unix ./var/run/secrets/workload-spiffe-uds/socket: bind: no such file or directory
2022-12-29T08:09:05.522196Z	warn	Failed to create directory for ./var/run/secrets/workload-spiffe-uds/socket: mkdir var/run/secrets: permission denied
2022-12-29T08:09:05.522323Z	error	sds	SDS grpc server for workload proxies failed to set up UDS: failed to listen on unix socket "./var/run/secrets/workload-spiffe-uds/socket": listen unix ./var/run/secrets/workload-spiffe-uds/socket: bind: no such file or directory
2022-12-29T08:09:21.523459Z	warn	sds	SDS grpc server could not be started
2022-12-29T08:36:56.952157Z	warn	failed to fetch token from file: open ./var/run/secrets/tokens/istio-token: no such file or directory
2022-12-29T08:36:56.952399Z	info	xdsproxy	connected to upstream XDS server: istiod.istio-system.svc:15012
2022-12-29T09:05:45.822787Z	warn	failed to fetch token from file: open ./var/run/secrets/tokens/istio-token: no such file or directory
2022-12-29T09:05:45.822906Z	info	xdsproxy	connected to upstream XDS server: istiod.istio-system.svc:15012

### Affected product area

- [ ] Ambient
- [X] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [X] Virtual Machine
- [ ] Control Plane Revisions

### Is this the right place to submit this?

- [X] This is not a security vulnerability
- [X] This is not a question about how to use Istio

**Comment by howardjohn** [comment]
Why do we not have permission to create folder, but can create files? Is it that typically the folder is created first then granted permission to istio-proxy?

The folder deletion is done by the OS btw, not Istio

**Comment by Christoph-Raab** [comment]
> Why do we not have permission to create folder, but can create files? Is it that typically the folder is created first then granted permission to istio-proxy?

Yes, from the [installation guide](https://istio.io/latest/docs/setup/install/virtual-machine/)

> 3. Install the token at /var/run/secrets/tokens:

```
$ sudo  mkdir -p /var/run/secrets/tokens
[...]
``` 

> 8. Transfer ownership of the files in /etc/certs/ and /var/lib/istio/envoy/ to the Istio proxy:

```
[...]
$ sudo chown -R istio-proxy [...] /var/run/secrets [...]
```

> The folder deletion is done by the OS btw, not Istio

Thanks for pointing that out. Any advice/best practice on how to recreate the folder after reboot? I would suggest to add it to the docs page linked above.

**Comment by s-johansson** [comment]
A feasible solution would be to add creation and setting permission to systemd unit file :
`ExecStart=/usr/bin/install -d -o istio-proxy -m 0755 /var/run/secrets`

**Comment by Christoph-Raab** [comment]
Solved with https://github.com/istio/istio/pull/43163
